Xrootd per site statistics

bbockelm opened the pull request:
```
Add new storage statistics service for Xrootd.

This service keeps per-site statistics from the Xrootd layer.
Previously, the IO statistics were kept by the StatisticsSenderService
on a per-file level.  These are less-useful now that Xrootd reads
from multiple servers (and possibly multiple sites) at a time!

Currently, this has to be enabled manually and only outputs
to the FJR.  Long-term, we plan on adding additional statistics
and additional output mechanisms (correcting the now-broken
StatisticsSenderService reports we send to the Dashboard).

NOTE: there is some trickery here due to the fact that the service
registry is only available on CMSSW-managed threads.  In order to
make the service available from the callback threads, the service
pointer is cached by the request manager object.
```

bbockelm committed:
Add new storage statistics service for Xrootd.

This service keeps per-site statistics from the Xrootd layer.
Previously, the IO statistics were kept by the StatisticsSenderService
on a per-file level.  These are less-useful now that Xrootd reads
from multiple servers (and possibly multiple sites) at a time!

Currently, this has to be enabled manually and only outputs
to the FJR.  Long-term, we plan on adding additional statistics
and additional output mechanisms (correcting the now-broken
StatisticsSenderService reports we send to the Dashboard).

NOTE: there is some trickery here due to the fact that the service
registry is only available on CMSSW-managed threads.  In order to
make the service available from the callback threads, the service
pointer is cached by the request manager object.

bbockelm committed:
Integrate the statistics service with the XrdAdaptor.

cmsbuild commented:
A new Pull Request was created by @bbockelm (Brian Bockelman) for CMSSW_7_4_X.

Xrootd per site statistics

It involves the following packages:

Utilities/XrdAdaptor


@cmsbuild, @Dr15Jones, @ktf, @nclopezo can you please review it and eventually sign? Thanks.
@Martin-Grunewald, @wddgit this is something you requested to watch as well.
You can sign-off by replying to this message having '+1' in the first line of your reply.
You can reject by replying  to this message having '-1' in the first line of your reply.
If you are a L2 or a release manager you can ask for tests by saying 'please test' in the first line of a comment.

Dr15Jones commented on Utilities/XrdAdaptor/src/XrdStatistics.cc:
```diff
@@ -0,0 +1,147 @@
+{
+    edm::ParameterSetDescription desc;
+    desc.setComment("Report Xrootd-related statistics centrally.");
+    desc.addUntracked<bool>("reportToFJR", true)
```
There is no such parameter used in the code

bbockelm replied:
Shoot, I forgot to hook that up.  Will update that in the next commit or so.

Dr15Jones commented on Utilities/XrdAdaptor/src/XrdRequestManager.cc:
```diff
@@ -107,6 +109,8 @@ RequestManager::RequestManager(const std::string &filename, XrdCl::OpenFlags::Fl
       m_excluded_active_count(0)
 {
+    edm::Service<XrdStatisticsService> statsService;
+    if (statsService.isAvailable()) {m_stats = &(*statsService);}
```
Since you are worried about service availability on a given thread maybe you should instead
-create a singleton
-have the XrdStaticstsService read from that singleton (or initialize the singleton as well?)
It appears the only reason you created a service was to listen to the 'endJob' transition.

bbockelm replied:
Ok, will do!

bbockelm committed:
Address issues from code review.

In particular,
- We actually use the pset description to determine whether stats
  are kept.
- State information is kept in a singleton, rather than passing
  a service pointr between objects.

cmsbuild commented:
+1
Tested at: 76d430dafef1799d51a6764a18c3991db110d740
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-7056/1804/summary.html

cmsbuild commented:
Pull request #7056 was updated. @Dr15Jones, @ktf can you please check and sign again.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-7056/1804/summary.html

Dr15Jones commented on Utilities/XrdAdaptor/src/XrdStatistics.cc:
```diff
@@ -0,0 +1,177 @@
+void
+XrdSiteStatisticsInformation::createInstance()
+{
+    std::lock_guard<std::mutex> lock(m_mutex);
```
I don't believe there is a need for the ```m_mutex``` since making multiple copies of ```XrdSiteStaticistsInformation``` doesn't appear to be a problem. So I suggest that the code be
```
if(not m_instance) {
   std::unique_ptr<XrdSiteStatisticsInformation> tmp { new XrdSiteStatisticsInformation() };
   XrdSiteStatisticsInformation* expected = nullptr;
   if( m_instance.compare_exchange_strong(expected,tmp.get()) {
      //this thread succeeded in the exchange so we need to no longer own the memory
      tmp.release();
   }
}
```

bbockelm replied:
The constructor calls edm::ActivityRegistry::watchPostEndJob.  Do I need to worry about pointer lifetime issues then (i.e., I don't see a way to "un-register")?

Dr15Jones replied:
The ```XrdSiteStatisticsService``` is guaranteed to exist beyond all calls the ```ActivityRegistry``` will make. Services are actually the last loadable components the framework deletes as it is shutting down.

bbockelm replied:
Ah - I see what got me confused.  Nevermind, your suggested change will work just fine.

Dr15Jones commented on Utilities/XrdAdaptor/src/XrdSource.cc:
```diff
@@ -103,6 +105,11 @@ Source::Source(timespec now, std::unique_ptr<XrdCl::File> fh, const std::string
+    XrdSiteStatisticsInformation *statsService = XrdSiteStatisticsInformation::getInstance();
+    if (statsService)
+    {
+        m_stats = statsService->getStatisticsForSite(m_site);
```
It looks to me like ```getStatisticsForSite``` is not thread safe. Are ```Source```s ever created on different threads?

Dr15Jones commented:
@bbockelm ping?

bbockelm commented:
Hi Chris,

Yup - this one is still on my docket.  I just have to service a few non-EDM-related interrupts...  In particular, I found one place where we invoke the XrdAdaptor::Source constructor from a threaded context, so I think I need to at least make getStatisticsForSite thread-safe.

Brian

Dr15Jones commented:
@bbockelm ping?

bbockelm committed:
Fixes from code review.

In particular, get rid of mutex when constructing the object; instead
use an atomic for synchronization.  *Do* use a mutex when creating
site statistics information as there are some obscure cases when
XrdAdaptor::Source can be created from multiple threads.

davidlange6 commented:
@cmsbuild, please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
Pull request #7056 was updated. @cmsbuild, @Dr15Jones, @ktf, @nclopezo can you please check and sign again.

davidlange6 commented:
@Dr15Jones - looks like there are commits here..

cmsbuild commented:
+1
Tested at: f1ea7c6993d1e19f58293e3e6ef79975e9a8c87e
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-7056/2445/summary.html

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-7056/2445/summary.html

Dr15Jones commented on Utilities/XrdAdaptor/src/XrdStatistics.cc:
```diff
@@ -0,0 +1,179 @@
+
+    for (std::shared_ptr<XrdSiteStatistics> const &stats : instance->m_sites)
+    {
+        stats->recomputeProperties();
```
```recomputeProperties``` is the only thread unsafe function in this class. Given you compute and then use the values immediately, I see no reason for them to be stored in a member data. Instead, pass in your ```std::map``` as an argument to ```computeProperties``` and then pass that map to ```reportPerformanceForModule```. It will be just as efficient but make the ```XrdSiteStatistics``` fully thread safe.

bbockelm committed:
Make rest of XrdSiteStatistics thread-safe.

Dr15Jones commented on Utilities/XrdAdaptor/src/XrdStatistics.h:
```diff
@@ -0,0 +1,116 @@
+
+    void postEndJob();
+
+    void fillDescriptions(edm::ConfigurationDescriptions &descriptions);
```
This has to be 'static' in order for it to ever be called.

Dr15Jones commented on Utilities/XrdAdaptor/src/XrdStatistics.cc:
```diff
@@ -0,0 +1,180 @@
+
+
+void
+XrdSiteStatistics::recomputeProperties(std::map<std::string, std::string> &props)
```
In the future, this should really be made ```const```. I won't require it for this pull request.

Dr15Jones commented:
+1

Dr15Jones commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next CMSSW_7_4_X IBs unless changes or unless it breaks tests. This pull request requires discussion in the ORP meeting before it's merged. @davidlange6, @degano, @ktf, @smuzaffar

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next CMSSW_7_4_X IBs unless changes or unless it breaks tests. This pull request requires discussion in the ORP meeting before it's merged. @davidlange6, @degano, @ktf, @smuzaffar

cmsbuild commented:
+1
Tested at: 67357c0e967a319716aa671d512eafc7706570e0
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-7056/2479/summary.html

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next CMSSW_7_4_X IBs unless changes (tests are also fine). This pull request requires discussion in the ORP meeting before it's merged. @davidlange6, @degano, @ktf, @smuzaffar

davidlange6 commented:
+1
